db/version_set.cc

```diff
@@ -591,14 +597,52 @@ class LevelIterator final : public InternalIterator {
   const std::vector<AtomicCompactionUnitBoundary>* compaction_boundaries_;
 };
 
+// void LevelIterator::Seek(const Slice& target) {
+//   size_t new_file_index = FindFile(icomparator_, *flevel_, target);
+
+//   InitFileIterator(new_file_index);
+//   if (file_iter_.iter() != nullptr) {
+//     file_iter_.Seek(target);
+//   }
+//   SkipEmptyFileForward();
+// }
+
 void LevelIterator::Seek(const Slice& target) {
+  bool reach_bound = false;
   size_t new_file_index = FindFile(icomparator_, *flevel_, target);
+  while (new_file_index < flevel_->num_files) {
+    auto file = flevel_->files[new_file_index];
+    // prune files by upper bound
+    if (read_options_.iterate_upper_bound && icomparator_.user_comparator()->Compare(file.smallest_key, *read_options_.iterate_upper_bound) >= 0) {
```

Comment:
Not 100% sure but I think `file.smallest_key` is an internal key. You need to wrap it with `ExtractUserKey` like `ExtractUserKey(file.smallest_key)`.